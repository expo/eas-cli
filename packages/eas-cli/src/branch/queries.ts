import chalk from 'chalk';
import CliTable from 'cli-table3';

import { PaginatedQueryOptions } from '../commandUtils/pagination';
import { UpdateBranchFragment } from '../graphql/generated';
import { BranchQuery } from '../graphql/queries/BranchQuery';
import Log from '../log';
import { UPDATE_COLUMNS, formatUpdateMessage, getPlatformsForGroup } from '../update/utils';
import { printJsonOnlyOutput } from '../utils/json';
import {
  paginatedQueryWithConfirmPromptAsync,
  paginatedQueryWithSelectPromptAsync,
} from '../utils/queries';

export const BRANCHES_LIMIT = 50;

export async function selectBranchOnAppAsync({
  projectId,
  promptTitle,
  displayTextForListItem,
  paginatedQueryOptions,
}: {
  projectId: string;
  displayTextForListItem: (queryItem: UpdateBranchFragment) => string;
  promptTitle: string;
  paginatedQueryOptions: PaginatedQueryOptions;
}): Promise<UpdateBranchFragment> {
  if (paginatedQueryOptions.nonInteractive) {
    throw new Error('Unable to select a branch in non-interactive mode.');
  }

  const selectedBranch = await paginatedQueryWithSelectPromptAsync({
    limit: paginatedQueryOptions.limit ?? BRANCHES_LIMIT,
    offset: paginatedQueryOptions.offset,
    queryToPerform: (limit, offset) => queryBranchesOnProjectAsync(limit, offset, projectId),
    promptOptions: {
      title: promptTitle,
      getIdentifierForQueryItem: updateBranchFragment => updateBranchFragment.id,
      createDisplayTextForSelectionPromptListItem: displayTextForListItem,
    },
  });
  if (!selectedBranch) {
    throw new Error(`Could not find any branches for project "${projectId}"`);
  }
  return selectedBranch;
}

export async function listAndRenderBranchesOnAppAsync({
  projectId,
  paginatedQueryOptions,
}: {
  projectId: string;
  paginatedQueryOptions: PaginatedQueryOptions;
}): Promise<void> {
  if (paginatedQueryOptions.nonInteractive) {
    const branches = await queryBranchesOnProjectAsync(
      paginatedQueryOptions.limit ?? BRANCHES_LIMIT,
      paginatedQueryOptions.offset,
      projectId
    );
    renderPageOfBranches(branches, paginatedQueryOptions);
  } else {
    await paginatedQueryWithConfirmPromptAsync({
      limit: paginatedQueryOptions.limit ?? BRANCHES_LIMIT,
      offset: paginatedQueryOptions.offset,
      queryToPerform: (limit, offset) => queryBranchesOnProjectAsync(limit, offset, projectId),
      promptOptions: {
        title: 'Load more branches?',
        renderListItems: branches => renderPageOfBranches(branches, paginatedQueryOptions),
      },
    });
  }
}

async function queryBranchesOnProjectAsync(
  limit: number,
  offset: number,
  projectId: string
): Promise<UpdateBranchFragment[]> {
  return await BranchQuery.listBranchesOnAppAsync({
    appId: projectId,
    limit,
    offset,
  });
}

function renderPageOfBranches(
  currentPage: UpdateBranchFragment[],
  { json }: PaginatedQueryOptions
): void {
  if (json) {
    printJsonOnlyOutput(currentPage);
  } else {
    const table = new CliTable({
      head: ['Branch', ...UPDATE_COLUMNS],
      wordWrap: true,
    });

    table.push(
      ...currentPage.map(branch => {
        if (branch.updates.length === 0) {
          return [branch.name, 'N/A', 'N/A', 'N/A', 'N/A'];
        }

        const latestUpdateOnBranch = branch.updates[0];
        return [
          branch.name,
          formatUpdateMessage(latestUpdateOnBranch),
          latestUpdateOnBranch.runtimeVersion,
          latestUpdateOnBranch.group,
          getPlatformsForGroup({
            group: latestUpdateOnBranch.group,
            updates: branch.updates,
          }),
        ];
      })
    );

    Log.addNewLineIfNone();
    Log.log(chalk.bold('Branches:'));
    Log.addNewLineIfNone();
    Log.log(table.toString());
  }
}
