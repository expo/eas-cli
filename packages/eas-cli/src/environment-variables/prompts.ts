import { EnvironmentVariableEnvironment } from '../graphql/generated';
import { promptAsync, selectAsync } from '../prompts';

export async function promptVariableEnvironmentAsync(
  nonInteractive: boolean
): Promise<EnvironmentVariableEnvironment> {
  if (nonInteractive) {
    throw new Error('Environment may not be empty.');
  }
  return await selectAsync('Select environment:', [
    { title: 'Development', value: EnvironmentVariableEnvironment.Development },
    { title: 'Preview', value: EnvironmentVariableEnvironment.Preview },
    { title: 'Production', value: EnvironmentVariableEnvironment.Production },
  ]);
}
export async function promptVariableValueAsync(nonInteractive: boolean): Promise<string> {
  const validationMessage = 'Variable value may not be empty.';
  if (nonInteractive) {
    throw new Error(validationMessage);
  }

  const { variableValue } = await promptAsync({
    type: 'text',
    name: 'variableValue',
    message: 'Variable value:',
    // eslint-disable-next-line async-protect/async-suffix
    validate: async variableValue => {
      if (!variableValue) {
        return validationMessage;
      }
      return true;
    },
  });

  if (!variableValue) {
    throw new Error(validationMessage);
  }

  return variableValue;
}
export async function promptVariableNameAsync(nonInteractive: boolean): Promise<string> {
  const validationMessage = 'Variable name may not be empty.';
  if (nonInteractive) {
    throw new Error(validationMessage);
  }

  const { name } = await promptAsync({
    type: 'text',
    name: 'name',
    message: `Variable name:`,
    validate: value => {
      if (!value) {
        return validationMessage;
      }

      if (!value.match(/^\w+$/)) {
        return 'Names may contain only letters, numbers, and underscores.';
      }

      return true;
    },
  });

  if (!name) {
    throw new Error(validationMessage);
  }

  return name;
}
