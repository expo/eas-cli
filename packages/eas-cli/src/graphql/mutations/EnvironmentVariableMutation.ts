import { print } from 'graphql';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../../commandUtils/context/contextUtils/createGraphqlClient';
import { withErrorHandlingAsync } from '../client';
import {
  CreateEnvironmentVariableForAccountMutation,
  CreateEnvironmentVariableForAppMutation,
  DeleteEnvironmentVariableMutation,
  EnvironmentVariableFragment,
  EnvironmentVariableVisibility,
  LinkSharedEnvironmentVariableMutation,
  UnlinkSharedEnvironmentVariableMutation,
} from '../generated';
import { EnvironmentVariableFragmentNode } from '../types/EnvironmentVariable';

type UpdateVariableArgs = {
  value?: string;
  name: string;
  overwrite: true;
  visibility?: EnvironmentVariableVisibility;
};

export type EnvironmentVariablePushInput = {
  name: string;
  value: string;
  environment: string;
  visibility: EnvironmentVariableVisibility;
  overwrite?: boolean;
};

export const EnvironmentVariableMutation = {
  async linkSharedEnvironmentVariableAsync(
    graphqlClient: ExpoGraphqlClient,
    environmentVariableId: string,
    appId: string,
    environment: string
  ): Promise<EnvironmentVariableFragment> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .mutation<LinkSharedEnvironmentVariableMutation>(
          gql`
            mutation LinkSharedEnvironmentVariable(
              $appId: ID!
              $environment: EnvironmentVariableEnvironment!
              $environmentVariableId: ID!
            ) {
              environmentVariable {
                linkSharedEnvironmentVariable(
                  appId: $appId
                  environmentVariableId: $environmentVariableId
                  environment: $environment
                ) {
                  id
                  ...EnvironmentVariableFragment
                }
              }
            }
            ${print(EnvironmentVariableFragmentNode)}
          `,
          { appId, environment, environmentVariableId }
        )
        .toPromise()
    );

    return data.environmentVariable.linkSharedEnvironmentVariable;
  },
  async unlinkSharedEnvironmentVariableAsync(
    graphqlClient: ExpoGraphqlClient,
    environmentVariableId: string,
    appId: string,
    environment: string
  ): Promise<EnvironmentVariableFragment> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .mutation<UnlinkSharedEnvironmentVariableMutation>(
          gql`
            mutation UnlinkSharedEnvironmentVariable(
              $appId: ID!
              $environment: EnvironmentVariableEnvironment!
              $environmentVariableId: ID!
            ) {
              environmentVariable {
                unlinkSharedEnvironmentVariable(
                  appId: $appId
                  environmentVariableId: $environmentVariableId
                  environment: $environment
                ) {
                  id
                  ...EnvironmentVariableFragment
                }
              }
            }
            ${print(EnvironmentVariableFragmentNode)}
          `,
          { appId, environment, environmentVariableId }
        )
        .toPromise()
    );

    return data.environmentVariable.unlinkSharedEnvironmentVariable;
  },
  async createSharedVariableAsync(
    graphqlClient: ExpoGraphqlClient,
    input:
      | {
          name: string;
          value: string;
          visibility: EnvironmentVariableVisibility;
          overwrite?: boolean;
        }
      | UpdateVariableArgs,
    accountId: string
  ): Promise<EnvironmentVariableFragment> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .mutation<CreateEnvironmentVariableForAccountMutation>(
          gql`
            mutation CreateEnvironmentVariableForAccount(
              $input: CreateSharedEnvironmentVariableInput!
              $accountId: ID!
            ) {
              environmentVariable {
                createEnvironmentVariableForAccount(
                  environmentVariableData: $input
                  accountId: $accountId
                ) {
                  id
                  ...EnvironmentVariableFragment
                }
              }
            }
            ${print(EnvironmentVariableFragmentNode)}
          `,
          { input, accountId }
        )
        .toPromise()
    );

    return data.environmentVariable.createEnvironmentVariableForAccount;
  },
  async createForAppAsync(
    graphqlClient: ExpoGraphqlClient,
    input:
      | {
          name: string;
          value?: string;
          environment: string;
          visibility: EnvironmentVariableVisibility;
          overwrite?: boolean;
        }
      | (UpdateVariableArgs & { environment: string }),
    appId: string
  ): Promise<EnvironmentVariableFragment> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .mutation<CreateEnvironmentVariableForAppMutation>(
          gql`
            mutation CreateEnvironmentVariableForApp(
              $input: CreateEnvironmentVariableInput!
              $appId: ID!
            ) {
              environmentVariable {
                createEnvironmentVariableForApp(environmentVariableData: $input, appId: $appId) {
                  id
                  ...EnvironmentVariableFragment
                }
              }
            }
            ${print(EnvironmentVariableFragmentNode)}
          `,
          { input, appId }
        )
        .toPromise()
    );

    return data.environmentVariable.createEnvironmentVariableForApp;
  },
  async deleteAsync(graphqlClient: ExpoGraphqlClient, id: string): Promise<{ id: string }> {
    const data = await withErrorHandlingAsync(
      graphqlClient
        .mutation<DeleteEnvironmentVariableMutation>(
          gql`
            mutation DeleteEnvironmentVariable($id: ID!) {
              environmentVariable {
                deleteEnvironmentVariable(id: $id) {
                  id
                }
              }
            }
          `,
          { id }
        )
        .toPromise()
    );

    return data.environmentVariable.deleteEnvironmentVariable;
  },
};
