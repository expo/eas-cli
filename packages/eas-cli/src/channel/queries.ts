import chalk from 'chalk';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../commandUtils/context/contextUtils/createGraphqlClient';
import { PaginatedQueryOptions } from '../commandUtils/pagination';
import { withErrorHandlingAsync } from '../graphql/client';
import {
  CreateUpdateChannelOnAppMutation,
  CreateUpdateChannelOnAppMutationVariables,
  ViewBranchesOnUpdateChannelQueryVariables,
  ViewUpdateChannelsOnAppQueryVariables,
} from '../graphql/generated';
import { BranchQuery, UpdateBranchOnChannelObject } from '../graphql/queries/BranchQuery';
import { ChannelQuery, UpdateChannelObject } from '../graphql/queries/ChannelQuery';
import Log from '../log';
import formatFields from '../utils/formatFields';
import { printJsonOnlyOutput } from '../utils/json';
import {
  paginatedQueryWithConfirmPromptAsync,
  paginatedQueryWithSelectPromptAsync,
} from '../utils/queries';
import { logChannelDetails } from './utils';

export const CHANNELS_LIMIT = 25;

export async function selectChannelOnAppAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    projectId,
    selectionPromptTitle,
    paginatedQueryOptions,
  }: {
    projectId: string;
    selectionPromptTitle: string;
    paginatedQueryOptions: PaginatedQueryOptions;
  }
): Promise<UpdateChannelObject> {
  if (paginatedQueryOptions.nonInteractive) {
    throw new Error('Unable to select a channel in non-interactive mode.');
  }

  const updateChannel = await paginatedQueryWithSelectPromptAsync({
    limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
    offset: paginatedQueryOptions.offset,
    queryToPerform: (limit, offset) =>
      queryChannelsOnAppAsync(graphqlClient, { appId: projectId, limit, offset }),
    promptOptions: {
      title: selectionPromptTitle,
      makePartialChoiceObject: updateChannel => ({ title: updateChannel.name }),
      getIdentifierForQueryItem: updateChannel => updateChannel.id,
    },
  });

  if (!updateChannel) {
    throw new Error(`Could not find any channels for app "${projectId}"`);
  }
  return updateChannel;
}

export async function listAndRenderChannelsOnAppAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    projectId,
    paginatedQueryOptions,
  }: {
    projectId: string;
    paginatedQueryOptions: PaginatedQueryOptions;
  }
): Promise<void> {
  if (paginatedQueryOptions.nonInteractive) {
    const channels = await queryChannelsOnAppAsync(graphqlClient, {
      appId: projectId,
      limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
      offset: paginatedQueryOptions.offset,
    });
    renderPageOfChannels(channels, paginatedQueryOptions);
  } else {
    await paginatedQueryWithConfirmPromptAsync({
      limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
      offset: paginatedQueryOptions.offset,
      queryToPerform: (limit, offset) =>
        queryChannelsOnAppAsync(graphqlClient, { limit, offset, appId: projectId }),
      promptOptions: {
        title: 'Load more channels?',
        renderListItems: channels => renderPageOfChannels(channels, paginatedQueryOptions),
      },
    });
  }
}

export async function listAndRenderBranchesAndUpdatesOnChannelAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    projectId: appId,
    channelName,
    paginatedQueryOptions,
  }: {
    projectId: string;
    channelName: string;
    paginatedQueryOptions: PaginatedQueryOptions;
  }
): Promise<void> {
  const channel = await ChannelQuery.viewUpdateChannelAsync(graphqlClient, { appId, channelName });
  renderChannelHeaderContent({ channelName: channel.name, channelId: channel.id });

  if (paginatedQueryOptions.nonInteractive) {
    const branches = await queryBranchesAndUpdateGroupsOnChannelAsync(graphqlClient, {
      appId,
      channelName,
      offset: paginatedQueryOptions.offset,
      limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
    });
    renderPageOfBranchesOnChannel(channel, branches, paginatedQueryOptions);
  } else {
    await paginatedQueryWithConfirmPromptAsync({
      limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
      offset: paginatedQueryOptions.offset,
      queryToPerform: (limit, offset) =>
        queryBranchesAndUpdateGroupsOnChannelAsync(graphqlClient, {
          channelName,
          appId,
          offset,
          limit,
        }),
      promptOptions: {
        title: 'Load more channels?',
        renderListItems: branches =>
          renderPageOfBranchesOnChannel(channel, branches, paginatedQueryOptions),
      },
    });
  }
}

async function queryChannelsOnAppAsync(
  graphqlClient: ExpoGraphqlClient,
  { appId, offset, limit }: ViewUpdateChannelsOnAppQueryVariables
): Promise<UpdateChannelObject[]> {
  return await ChannelQuery.viewUpdateChannelsOnAppAsync(graphqlClient, {
    appId,
    offset,
    limit,
  });
}

async function queryBranchesAndUpdateGroupsOnChannelAsync(
  graphqlClient: ExpoGraphqlClient,
  args: ViewBranchesOnUpdateChannelQueryVariables
): Promise<UpdateBranchOnChannelObject[]> {
  return await BranchQuery.listBranchesOnChannelAsync(graphqlClient, args);
}

function renderPageOfChannels(
  currentPage: UpdateChannelObject[],
  { json }: PaginatedQueryOptions
): void {
  if (json) {
    printJsonOnlyOutput({ currentPage });
  } else {
    for (const channel of currentPage) {
      renderChannelHeaderContent({ channelName: channel.name, channelId: channel.id });
      Log.addNewLineIfNone();
      logChannelDetails(channel);

      if (currentPage.indexOf(channel) < currentPage.length - 1) {
        Log.log(`\n${chalk.dim('———')}\n`);
      }
    }
  }
}

function renderPageOfBranchesOnChannel(
  channel: UpdateChannelObject,
  currentPage: UpdateBranchOnChannelObject[],
  { json }: PaginatedQueryOptions
): void {
  const channelWithNewBranches = { ...channel, updateBranches: currentPage };
  if (json) {
    printJsonOnlyOutput({ currentPage: channelWithNewBranches });
  } else {
    // The channel details contain both the branch and latest update group
    Log.addNewLineIfNone();
    logChannelDetails(channelWithNewBranches);
  }
}

function renderChannelHeaderContent({
  channelName,
  channelId,
}: {
  channelName: string;
  channelId: string;
}): void {
  Log.addNewLineIfNone();
  Log.log(chalk.bold('Channel:'));
  Log.log(
    formatFields([
      { label: 'Name', value: channelName },
      { label: 'ID', value: channelId },
    ])
  );
  Log.addNewLineIfNone();
  Log.log(chalk`{bold Branches pointed at this channel and their most recent update group:}`);
}

export async function createChannelOnAppAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    appId,
    branchId,
    channelName,
  }: {
    appId: string;
    branchId: string;
    channelName: string;
  }
): Promise<CreateUpdateChannelOnAppMutation> {
  // Point the new channel at a branch with its same name.
  const branchMapping = JSON.stringify({
    data: [{ branchId, branchMappingLogic: 'true' }],
    version: 0,
  });
  return await withErrorHandlingAsync(
    graphqlClient
      .mutation<CreateUpdateChannelOnAppMutation, CreateUpdateChannelOnAppMutationVariables>(
        gql`
          mutation CreateUpdateChannelOnApp($appId: ID!, $name: String!, $branchMapping: String!) {
            updateChannel {
              createUpdateChannelForApp(appId: $appId, name: $name, branchMapping: $branchMapping) {
                id
                name
                branchMapping
              }
            }
          }
        `,
        {
          appId,
          name: channelName,
          branchMapping,
        }
      )
      .toPromise()
  );
}

export async function ensureChannelExistsAsync(
  graphqlClient: ExpoGraphqlClient,
  { appId, branchId, channelName }: { appId: string; branchId: string; channelName: string }
): Promise<void> {
  try {
    await createChannelOnAppAsync(graphqlClient, {
      appId,
      channelName,
      branchId,
    });
  } catch (e: any) {
    const isIgnorableError =
      e.graphQLErrors?.length === 1 &&
      e.graphQLErrors[0].extensions.errorCode === 'CHANNEL_ALREADY_EXISTS';
    if (!isIgnorableError) {
      throw e;
    }
  }
}
