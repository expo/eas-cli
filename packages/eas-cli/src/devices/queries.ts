import chalk from 'chalk';

import { ExpoGraphqlClient } from '../commandUtils/context/contextUtils/createGraphqlClient';
import { PaginatedQueryOptions } from '../commandUtils/pagination';
import { formatAppleTeam } from '../credentials/ios/actions/AppleTeamFormatting';
import { formatDeviceLabel } from '../credentials/ios/actions/DeviceUtils';
import { AppleDeviceQuery } from '../credentials/ios/api/graphql/queries/AppleDeviceQuery';
import { AppleTeamQuery } from '../credentials/ios/api/graphql/queries/AppleTeamQuery';
import { AppleDeviceFragment, AppleTeamFragment } from '../graphql/generated';
import Log from '../log';
import { printJsonOnlyOutput } from '../utils/json';
import {
  paginatedQueryWithConfirmPromptAsync,
  paginatedQueryWithSelectPromptAsync,
} from '../utils/queries';
import formatDevice, { AppleTeamIdAndName } from './utils/formatDevice';

export const TEAMS_LIMIT = 50;
export const DEVICES_LIMIT = 50;

export async function selectAppleTeamOnAccountAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    accountName,
    selectionPromptTitle,
    paginatedQueryOptions,
  }: {
    accountName: string;
    selectionPromptTitle: string;
    paginatedQueryOptions: PaginatedQueryOptions;
  }
): Promise<AppleTeamFragment> {
  if (paginatedQueryOptions.nonInteractive) {
    throw new Error('Unable to select an Apple team in non-interactive mode.');
  } else {
    const selectedAppleTeam = await paginatedQueryWithSelectPromptAsync({
      limit: paginatedQueryOptions.limit ?? TEAMS_LIMIT,
      offset: paginatedQueryOptions.offset,
      queryToPerform: (limit, offset) =>
        AppleTeamQuery.getAllForAccountAsync(graphqlClient, {
          accountName,
          limit,
          offset,
        }),
      promptOptions: {
        title: selectionPromptTitle,
        makePartialChoiceObject: appleTeam => ({
          title: appleTeam.appleTeamName
            ? `${appleTeam.appleTeamName} (ID: ${appleTeam.appleTeamIdentifier})`
            : appleTeam.appleTeamIdentifier,
        }),
        getIdentifierForQueryItem: appleTeam => appleTeam.id,
      },
    });
    if (!selectedAppleTeam) {
      throw new Error(`Couldn't find any teams for the account ${accountName}`);
    }
    return selectedAppleTeam;
  }
}

export async function selectAppleDeviceOnAppleTeamAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    accountName,
    appleTeamIdentifier,
    selectionPromptTitle,
    paginatedQueryOptions,
  }: {
    accountName: string;
    appleTeamIdentifier: string;
    selectionPromptTitle: string;
    paginatedQueryOptions: PaginatedQueryOptions;
  }
): Promise<AppleDeviceFragment> {
  if (paginatedQueryOptions.nonInteractive) {
    throw new Error('Unable to select an Apple device in non-interactive mode.');
  } else {
    const selectedAppleDevice = await paginatedQueryWithSelectPromptAsync({
      limit: paginatedQueryOptions.limit ?? DEVICES_LIMIT,
      offset: paginatedQueryOptions.offset,
      queryToPerform: (limit, offset) =>
        AppleDeviceQuery.getAllForAppleTeamAsync(graphqlClient, {
          accountName,
          appleTeamIdentifier,
          limit,
          offset,
        }),
      promptOptions: {
        title: selectionPromptTitle,
        makePartialChoiceObject: appleDevice => ({
          title: formatDeviceLabel(appleDevice),
        }),
        getIdentifierForQueryItem: appleTeam => appleTeam.id,
      },
    });
    if (!selectedAppleDevice) {
      throw new Error(
        `Couldn't find any devices on the Apple team with the id ${appleTeamIdentifier}`
      );
    }
    return selectedAppleDevice;
  }
}

export async function listAndRenderAppleDevicesOnAppleTeamAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    accountName,
    appleTeam,
    paginatedQueryOptions,
  }: {
    accountName: string;
    appleTeam: AppleTeamIdAndName;
    paginatedQueryOptions: PaginatedQueryOptions;
  }
): Promise<void> {
  if (paginatedQueryOptions.nonInteractive) {
    const devices = await AppleDeviceQuery.getAllForAppleTeamAsync(graphqlClient, {
      accountName,
      appleTeamIdentifier: appleTeam.appleTeamIdentifier,
      offset: paginatedQueryOptions.offset,
      limit: paginatedQueryOptions.limit,
    });
    renderPageOfAppleDevices({ devices, appleTeam, paginatedQueryOptions });
  } else {
    await paginatedQueryWithConfirmPromptAsync({
      limit: paginatedQueryOptions.limit ?? DEVICES_LIMIT,
      offset: paginatedQueryOptions.offset,
      queryToPerform: (limit, offset) =>
        AppleDeviceQuery.getAllForAppleTeamAsync(graphqlClient, {
          accountName,
          appleTeamIdentifier: appleTeam.appleTeamIdentifier,
          limit,
          offset,
        }),
      promptOptions: {
        title: 'Load more devices?',
        renderListItems: devices =>
          renderPageOfAppleDevices({ devices, appleTeam, paginatedQueryOptions }),
      },
    });
  }
}

function renderPageOfAppleDevices({
  devices,
  appleTeam,
  paginatedQueryOptions,
}: {
  devices: AppleDeviceFragment[];
  appleTeam: AppleTeamIdAndName;
  paginatedQueryOptions: PaginatedQueryOptions;
}): void {
  if (paginatedQueryOptions.json) {
    printJsonOnlyOutput(devices);
  } else {
    if (devices.length === 0) {
      Log.log(
        `Could not find devices on Apple team -- ${formatAppleTeam({
          appleTeamIdentifier: appleTeam.appleTeamIdentifier,
          appleTeamName: appleTeam.appleTeamName,
        })}`
      );
    } else {
      const list = devices
        .map(device => formatDevice(device, appleTeam))
        .join(`\n\n${chalk.dim('———')}\n\n`);

      Log.log(`\n${list}`);
      Log.addNewLineIfNone();
    }
  }
}
