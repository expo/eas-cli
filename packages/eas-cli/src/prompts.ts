import { constants } from 'os';
import prompts, { Answers, Choice, Options, PromptType, PromptObject as Question } from 'prompts';

export { PromptType, Question, Choice };

export interface ExpoChoice<T> extends Choice {
  value: T;
}

type NamelessQuestion = Omit<Question<'value'>, 'name' | 'type'>;

export async function promptAsync<T extends string = string>(
  questions: Question<T> | Question<T>[],
  options: Options = {}
): Promise<Answers<T>> {
  if (!process.stdin.isTTY && !global.test) {
    const message = Array.isArray(questions) ? questions[0]?.message : questions.message;
    throw new Error(
      `Input is required, but stdin is not readable. Failed to display prompt: ${message}`
    );
  }
  return await prompts<T>(questions, {
    onCancel() {
      process.exit(constants.signals.SIGINT + 128); // Exit code 130 used when process is interrupted with ctrl+c.
    },
    ...options,
  });
}

export async function confirmAsync(
  question: NamelessQuestion,
  options?: Options
): Promise<boolean> {
  const { value } = await promptAsync(
    {
      initial: true,
      ...question,
      name: 'value',
      type: 'confirm',
    },
    options
  );
  return value;
}

export async function selectAsync<T>(
  message: string,
  choices: ExpoChoice<T>[],
  config?: {
    options?: Options;
    warningMessageForDisabledEntries?: string;
  }
): Promise<T> {
  const { value } = await promptAsync(
    {
      message,
      choices,
      name: 'value',
      type: 'select',
      warn: config?.warningMessageForDisabledEntries,
    },
    config?.options ?? {}
  );
  return value ?? null;
}

/**
 * Create a more dynamic yes/no confirmation that can be cancelled.
 *
 * @param questions
 * @param options
 */
export async function toggleConfirmAsync(
  questions: NamelessQuestion,
  options?: Options
): Promise<boolean> {
  const { value } = await promptAsync(
    {
      active: 'yes',
      inactive: 'no',
      ...questions,
      name: 'value',
      type: 'toggle',
    },
    options
  );
  return value ?? null;
}

export async function pressAnyKeyToContinueAsync(): Promise<void> {
  process.stdin.setRawMode(true);
  process.stdin.resume();
  process.stdin.setEncoding('utf8');

  await new Promise<void>(res => {
    process.stdin.on('data', key => {
      if (String(key) === '\u0003') {
        process.exit(constants.signals.SIGINT + 128); // ctrl-c
      }
      res();
    });
  });
}
