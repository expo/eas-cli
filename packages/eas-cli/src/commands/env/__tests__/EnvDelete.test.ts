import { Config } from '@oclif/core';

import {
  EnvironmentVariableEnvironment,
  EnvironmentVariableScope,
} from '../../../graphql/generated';
import { EnvironmentVariableMutation } from '../../../graphql/mutations/EnvironmentVariableMutation';
import { EnvironmentVariablesQuery } from '../../../graphql/queries/EnvironmentVariablesQuery';
import Log from '../../../log';
import { promptAsync, toggleConfirmAsync } from '../../../prompts';
import EnvDelete from '../delete';

jest.mock('../../../graphql/queries/EnvironmentVariablesQuery');
jest.mock('../../../graphql/mutations/EnvironmentVariableMutation');
jest.mock('../../../prompts');
jest.mock('../../../log');

describe(EnvDelete, () => {
  const projectId = 'test-project-id';
  const variableId = '1';
  const graphqlClient = {};
  const mockConfig = {} as unknown as Config;
  const mockContext = {
    privateProjectConfig: { projectId },
    loggedIn: { graphqlClient },
  };

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('deletes a variable by name in non-interactive mode', async () => {
    const mockVariables = [
      {
        id: variableId,
        name: 'TEST_VARIABLE',
        scope: EnvironmentVariableScope.Project,
        environments: [EnvironmentVariableEnvironment.Production],
      },
    ];
    (EnvironmentVariablesQuery.byAppIdAsync as jest.Mock).mockResolvedValue(mockVariables);

    const command = new EnvDelete(
      [
        '--variable-name',
        'TEST_VARIABLE',
        '--variable-environment',
        'production',
        '--non-interactive',
      ],
      mockConfig
    );
    // @ts-expect-error
    jest.spyOn(command, 'getContextAsync').mockReturnValue(mockContext);
    await command.runAsync();

    expect(EnvironmentVariableMutation.deleteAsync).toHaveBeenCalledWith(graphqlClient, variableId);
    expect(Log.withTick).toHaveBeenCalledWith('️Deleted variable TEST_VARIABLE".');
  });

  it('prompts for variable selection when name is not provided', async () => {
    const mockVariables = [
      { id: variableId, name: 'TEST_VARIABLE', scope: EnvironmentVariableScope.Project },
    ];
    (EnvironmentVariablesQuery.byAppIdAsync as jest.Mock).mockResolvedValue(mockVariables);
    (promptAsync as jest.Mock).mockResolvedValue({ variable: mockVariables[0] });
    (toggleConfirmAsync as jest.Mock).mockResolvedValue(true);

    const command = new EnvDelete([], mockConfig);
    // @ts-expect-error
    jest.spyOn(command, 'getContextAsync').mockReturnValue(mockContext);
    await command.runAsync();

    expect(promptAsync).toHaveBeenCalled();
    expect(EnvironmentVariableMutation.deleteAsync).toHaveBeenCalledWith(graphqlClient, variableId);
    expect(Log.withTick).toHaveBeenCalledWith('️Deleted variable TEST_VARIABLE".');
  });

  it('throws an error when variable name is not found', async () => {
    const mockVariables = [
      { id: variableId, name: 'TEST_VARIABLE', scope: EnvironmentVariableScope.Project },
    ];
    (EnvironmentVariablesQuery.byAppIdAsync as jest.Mock).mockResolvedValue(mockVariables);

    const command = new EnvDelete(['--variable-name', 'NON_EXISTENT_VARIABLE'], mockConfig);

    // @ts-expect-error
    jest.spyOn(command, 'getContextAsync').mockReturnValue(mockContext);
    await expect(command.runAsync()).rejects.toThrow('Variable "NON_EXISTENT_VARIABLE" not found.');
  });

  it('throws an error when multiple variables with the same name are found', async () => {
    const mockVariables = [
      { id: variableId, name: 'TEST_VARIABLE', scope: EnvironmentVariableScope.Project },
      { id: '2', name: 'TEST_VARIABLE', scope: EnvironmentVariableScope.Project },
    ];
    (EnvironmentVariablesQuery.byAppIdAsync as jest.Mock).mockResolvedValue(mockVariables);

    const command = new EnvDelete(['--variable-name', 'TEST_VARIABLE'], mockConfig);

    // @ts-expect-error
    jest.spyOn(command, 'getContextAsync').mockReturnValue(mockContext);
    await expect(command.runAsync()).rejects.toThrow(
      'Multiple variables with name "TEST_VARIABLE" found. Please select the variable to delete interactively or run command with --variable-environment ENVIRONMENT option.'
    );
  });

  it('cancels deletion when user does not confirm', async () => {
    const mockVariables = [
      { id: variableId, name: 'TEST_VARIABLE', scope: EnvironmentVariableScope.Project },
    ];
    (EnvironmentVariablesQuery.byAppIdAsync as jest.Mock).mockResolvedValue(mockVariables);
    (promptAsync as jest.Mock).mockResolvedValue({ variable: mockVariables[0] });
    (toggleConfirmAsync as jest.Mock).mockResolvedValue(false);

    const command = new EnvDelete(['--non-interactive'], mockConfig);

    // @ts-expect-error
    jest.spyOn(command, 'getContextAsync').mockReturnValue(mockContext);
    await expect(command.runAsync()).rejects.toThrowErrorMatchingSnapshot();

    expect(EnvironmentVariableMutation.deleteAsync).not.toHaveBeenCalled();
  });
});
