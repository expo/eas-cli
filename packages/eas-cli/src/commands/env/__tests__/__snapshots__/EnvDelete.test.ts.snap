// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvDelete cancels deletion when user does not confirm 1`] = `"Environment variable needs 'name' to be specified when running in non-interactive mode. Run the command with [1m--variable-name VARIABLE_NAME[22m flag to fix the issue"`;
